### ISSUE TITLE
feat: three-notebook thesis figure and graph generation pipeline

### ISSUE THREAD

**Issue by SakifKhan98** [SakifKhan98/cattle-vision-framework#28]
## Problem Statement

The thesis requires two categories of output that currently have no unified generation pipeline:

1. **Quantitative charts** — training curves, F1 comparisons, perturbation heatmaps, tracking metrics, activity budgets, transition matrices — are spread across six ad-hoc notebooks (01–06) and a standalone plotting script. None target the full result set, none share a consistent visual style, and none save high-res publication-ready outputs.

2. **Visual figure plates** — sample frames from all six datasets annotated with detection boxes, segmentation masks, tracking IDs, behavior labels, and all ten perturbation conditions — have no generation pipeline. 92 segmentation samples and 3 behavior videos exist but cover only CBVD-5/CVB and are not composited into thesis plates.

The result is that thesis figure generation requires significant manual effort and produces inconsistent outputs. A structured, repeatable pipeline is needed before thesis writing begins.

## Solution

Three Jupyter notebooks that form a linear pipeline:

1. **`notebooks/graphs.ipynb`** — reads all committed CSVs and JSONs, produces every quantitative chart, saves to `results/figures/graphs/`. No GPU, fast reruns.
2. **`notebooks/render_figures.ipynb`** — selects representative frames, runs RF-DETR and SAM2 inference, renders four overlay types per frame per dataset plus all ten perturbation conditions, saves raw annotated PNGs and a manifest JSON to `results/figures/frames/`. Run once.
3. **`notebooks/assemble_plates.ipynb`** — reads the manifest, composites PNGs into publication-ready figure plates with captions, saves 300 DPI PNGs and paired `.tex` caption snippets to `results/figures/plates/`. Fast reruns for layout iteration.

Existing notebooks 01–06 are superseded by this pipeline and can be archived.

## User Stories

1. As a thesis author, I want training loss and macro-F1 curves for all five behavior configs (v1 and v2) in a single figure, so that I can show convergence behavior and compare in-domain vs. cross-domain training.
2. As a thesis author, I want a grouped bar chart of per-class F1 across all five configs and both model versions, so that I can discuss class-specific performance patterns.
3. As a thesis author, I want a cross-dataset mAP50 comparison bar chart covering all six datasets (CBVD-5, CVB, OpenCows2020, Cows2021, CattleEyeView, Freeman), so that I can visualize the OOD detection performance drop.
4. As a thesis author, I want a perturbation robustness heatmap (5 perturbation types × 6 datasets) at both low and high severity, annotated with delta_mAP50 values, so that I can show brightness as the dominant vulnerability.
5. As a thesis author, I want per-video MOTA and IDF1 box plots across 447 CVB videos, so that I can show tracking performance distribution and variance.
6. As a thesis author, I want a min_hits ablation bar chart showing IDF1, MOTA, and ID switches invariant across min_hits ∈ {1, 2, 3, 5}, so that I can justify the chosen hyperparameter.
7. As a thesis author, I want a stacked bar chart of activity budgets per dataset, so that I can show behavioral time allocation patterns.
8. As a thesis author, I want a behavior transition heatmap from the transition matrix CSV, so that I can show Lying↔Ruminating coupling and other transition patterns.
9. As a thesis author, I want a behavioral deviation scatter/box plot with outliers flagged, so that I can demonstrate the analytics layer's monitoring capability.
10. As a thesis author, I want a per-stage pipeline timing breakdown (horizontal stacked bar), so that I can discuss computational feasibility.
11. As a thesis author, I want segmentation coverage histograms for CBVD-5 and CVB, so that I can characterize mask completeness.
12. As a thesis author, I want all charts saved as high-res PNG with consistent fonts, colors, and axis styling, so that I can import them directly into LaTeX without reformatting.
13. As a thesis author, I want representative frames selected automatically (one per dominant behavior class per dataset), so that the visual figures cover behavioral diversity without manual curation.
14. As a thesis author, I want detection-only overlay frames for all six datasets, so that I can show RF-DETR performance on each domain.
15. As a thesis author, I want detection + segmentation mask overlay frames for all six datasets, so that I can illustrate the SAM2 segmentation stage.
16. As a thesis author, I want detection + segmentation + tracking ID overlay frames for CBVD-5 and CVB (where tracking JSONs exist), so that I can show identity-preserved multi-object tracking.
17. As a thesis author, I want full pipeline overlay frames (detection + segmentation + tracking ID + behavior label) for CBVD-5 and CVB, so that I can demonstrate the complete end-to-end system output.
18. As a thesis author, I want detection overlay frames rendered under all ten perturbation conditions (5 types × 2 severities) for all six datasets, so that I can visually demonstrate robustness degradation.
19. As a thesis author, I want a manifest JSON mapping every rendered frame to its file path by dataset, render type, and perturbation condition, so that the assembly notebook can load frames without hardcoded paths.
20. As a thesis author, I want a 6×4 dataset sample plate (6 datasets × 4 render types), so that I can show the full pipeline capability across all domains in one figure.
21. As a thesis author, I want a perturbation comparison plate per perturbation type (6 datasets × 2 severities), so that I can show how each perturbation affects each dataset.
22. As a thesis author, I want a perturbation severity strip for Freeman Center (clean + all 10 conditions), so that I can show progressive degradation in the best-performing OOD dataset.
23. As a thesis author, I want a full pipeline progression strip for CBVD-5 and CVB (det → det+seg → det+seg+track → full pipeline), so that I can walk through each pipeline stage visually.
24. As a thesis author, I want a behavior class example strip (3 frames per class × 7 classes) drawn from the tubelet dataset, so that I can illustrate what each recognized behavior looks like.
25. As a thesis author, I want all plates saved at 300 DPI with LaTeX-ready caption `.tex` snippets alongside each plate, so that I can include them in the thesis document without reformatting.

### Notebook 1 — `notebooks/graphs.ipynb`

- **Data sources (all committed, no inference needed):**
  - `results/behavior/training_logs/*.csv` — 11 files, columns: epoch, train_loss, val_loss, val_macro_f1, per-class F1, lr
  - `results/behavior/f1_per_class.csv` — 10 experiments, 7-class F1, precision, recall
  - `results/detection/*.json` + `results/generalization/ood_summary.csv` — per-dataset mAP50/75, AR
  - `results/generalization/perturbation_delta.csv` — 60 rows: dataset × perturbation_type × severity → delta_mAP50
  - `results/tracking/tracking_per_video_all.csv` — 447 rows: per-video MOTA, IDF1, MOTP, IDS
  - `results/tracking/minhits_ablation.csv` — 5 rows: min_hits → IDF1, MOTA, IDS
  - `results/analytics/activity_budget.csv`, `behavior_deviation.csv`, `transition_matrix.csv`
  - `results/inference/timing.csv`
  - `results/segmentation/{cbvd5,cvb}_segmentation_stats.csv`
- **Style:** seaborn-based, publication font sizes (title 14pt, axis labels 12pt, ticks 10pt), neutral palette with accent colors for datasets; consistent figure size 8×5 inches landscape
- **Output:** all PNGs to `results/figures/graphs/`, filenames match section slugs (e.g., `training_curves_cbvd5_v2.png`)
- **Existing overlap:** `notebooks/06_plot_charts.ipynb` covers Config A vs B RF-DETR-Seg training only — retain as-is, do not duplicate in graphs.ipynb; instead reference it in a note cell

### Notebook 2 — `notebooks/render_figures.ipynb`

- **Model loading:** RF-DETR checkpoint at `runs/detection/rfdetr_combined_v1/checkpoint_best_total.pth`; SAM2 at `weights/sam2.1_hiera_large.pt`; confidence threshold fixed at 0.3 (matches all prior evals)
- **Frame selection strategy:** for each dataset, pick 3–5 frames by scanning tracking JSONs (CBVD-5/CVB) or raw image directories (OOD datasets); prioritize frames where ≥2 cattle are visible and at least one dominant behavior class is present; save selection to `results/figures/frames/manifest.json`
- **Overlay rendering order:** detection only → add seg mask → add track ID → add behavior label. Each pass saves independently so partial runs are usable.
- **Tracking JSON coverage:** CBVD-5 and CVB have tracking JSONs in `data/processed/tracking_v2/`; OpenCows2020, Cows2021, CattleEyeView, Freeman do not — passes C and D for these datasets render detection+seg only, with track ID cell left blank
- **Behavior label source:** `results/behavior/predictions/{run}_val.csv` matched by tubelet_dir + frame range; falls back to "Unknown" if no match
- **Perturbation rendering:** albumentations transforms applied in-memory before inference, matching exact parameters used in `scripts/eval_detection_perturbation.py`; renders detection-only overlays (no seg/track) under perturbation to keep runtime manageable
- **Output structure:**
  ```
  results/figures/frames/
    manifest.json
    {dataset}/
      detection/
      det_seg/
      det_seg_track/       # cbvd5, cvb only
      full_pipeline/       # cbvd5, cvb only
      perturbations/
        {type}_{severity}/
  ```
- **Runtime estimate:** ~15–30 min for full run on RTX 3060; cells are checkpointed so already-rendered frames are skipped on re-run

### Notebook 3 — `notebooks/assemble_plates.ipynb`

- **Input:** `results/figures/frames/manifest.json` only — no direct file path hardcoding
- **Plate layout engine:** matplotlib GridSpec with PIL for image loading; uniform cell size with dataset/render-type axis labels
- **Caption format:** each plate gets a companion `{plate_name}_caption.tex` containing `\caption{...}\label{fig:{slug}}` ready for `\input{}` in LaTeX
- **Plate inventory:**
  1. `pipeline_overview_plate.png` — 6×4 grid (datasets × render types)
  2. `detection_gallery_{dataset}.png` — one per dataset, 1×N strip
  3. `perturbation_plate_{type}.png` — one per perturbation type, 6×2 grid
  4. `perturbation_severity_strip_freeman.png` — 1×11 (clean + 10 conditions)
  5. `pipeline_progression_{dataset}.png` — for cbvd5 and cvb
  6. `behavior_class_examples.png` — 7×3 grid from tubelets
- **300 DPI output** to `results/figures/plates/`

### Shared conventions

- `results/figures/` is **not gitignored** for plates and graphs (small PNGs); `results/figures/frames/` is gitignored (large intermediate renders)
- Both render_figures and assemble_plates notebooks include a "quick verify" cell at the top that checks manifest existence and prints a summary before running

## Testing Decisions

Jupyter notebooks are not unit-tested. Verification is by inspection:

- **graphs.ipynb:** each chart cell prints the source dataframe shape before plotting; a final cell lists all files written to `results/figures/graphs/` with file sizes
- **render_figures.ipynb:** manifest writer cell prints per-dataset frame counts and flags any dataset with zero frames selected; a "smoke test" cell renders one frame from CBVD-5 and displays inline before the full batch run
- **assemble_plates.ipynb:** each plate cell displays the plate inline at reduced resolution before saving; a final cell lists all plates written with pixel dimensions

No pytest tests are required for this work.

## Out of Scope

- Re-running detection/tracking/behavior inference to generate new predictions (all evaluation is already complete and committed)
- Generating figures for the RF-DETR-Seg (Config A vs B) experiment — already covered by `notebooks/06_plot_charts.ipynb`
- Video output (MP4) — thesis requires static PNG figures only
- Interactive/web figures (Plotly, Bokeh)
- Generating figures under the RF-DETR-Seg checkpoint (`seg_medium_lr5e5`) — the checkpoint re-run is deferred until after thesis writing is otherwise complete (see CLAUDE.md §10)
- LaTeX document generation — notebooks produce `.tex` caption snippets only, not full document structure

## Further Notes

- Existing notebooks 01–05 can be archived to `notebooks/_archive/` once the new pipeline is validated — they are superseded but should not be deleted until the new notebooks cover all their content
- The `results/figures/frames/` directory should be added to `.gitignore`; `results/figures/graphs/` and `results/figures/plates/` should be committed (outputs are small)
- Freeman Center raw data is video-only (54 AVI files, no extracted frames) — render_figures.ipynb must include a frame extraction step for this dataset before the overlay passes
- CVB raw data is frames-only (225,829 JPGs), no videos — frame selection reads directly from the image directory
